### PR TITLE
MicroOS: enable SELinux tests

### PR DESCRIPTION
**Pull request by jlausuch**
TEST SLE Micro and openSUSE MicroOS with SELinux enabled

- Related ticket: https://progress.opensuse.org/issues/91133

### CODE REVIEW

**Comment by jlausuch** [comment]
VRs:
- [MicroOS](https://openqa.opensuse.org/tests/1750717)
- [SLE Micro](http://fromm.arch.suse.de/tests/1181)

File: tests/transactional/enable_selinux.pm [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+#
+# Package: transactional-update
+# Summary: Disable Grub timeout by transactional update
+#
+# Maintainer: Jose Lausuch <jalausuch@suse.com>
```
Comment (b10n1k) [inline tests/transactional/enable_selinux.pm:13]: shouldnt be the qac mail here too?

File: tests/transactional/enable_selinux.pm [hunk 2]
```diff
@@ -0,0 +1,48 @@
+# SUSE's openQA tests
+#
+# Copyright © 2020 SUSE LLC
```
Comment (b10n1k) [inline tests/transactional/enable_selinux.pm:3]: s/2020/2021/

File: schedule/sle-micro/selinux.yaml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  maintenance:
+    FLAVOR:
+      'MicroOS-Image-Updates':
+        - transactional/install_updates
+schedule:
+  - microos/disk_boot
+  - transactional/disable_grub
+  - '{{registration}}'
+  - '{{maintenance}}'
+  - transactional/enable_selinux
```
Comment (b10n1k) [inline schedule/sle-micro/selinux.yaml:19]: why is it conditional in main.pm and is without.

File: products/microos/main.pm [hunk 4]
```diff
@@ -36,6 +36,7 @@ sub load_boot_from_disk_tests {
     # Preparation for start testing
     loadtest 'microos/disk_boot';
     loadtest 'installation/system_workarounds' if is_aarch64;
+    loadtest 'transactional/enable_selinux'    if (get_var("SELINUX"));
```
Comment (b10n1k) [inline products/microos/main.pm:39]: could you add `SELINUX` in the variables.pm?

File: tests/transactional/enable_selinux.pm (hunk 1 above)
Comment (jlausuch) [reply tests/transactional/enable_selinux.pm:13]: yes, sorry, I copy pasted

File: tests/transactional/enable_selinux.pm (hunk 2 above)
Comment (jlausuch) [reply tests/transactional/enable_selinux.pm:3]: copy paste :)

File: schedule/sle-micro/selinux.yaml (hunk 3 above)
Comment (jlausuch) [reply schedule/sle-micro/selinux.yaml:19]: you are right, this is not needed. The schedule is basically the same as for the image, I will just add the variable. Thanks

**Review by b10n1k — CHANGES_REQUESTED** [review]
I wonder if you add another module which will check for *denial* in the audit. wdyt?

**Comment by jlausuch** [comment]
> I wonder if you should add another module which will check for _denial_ in the audit. wdyt?

If we need to do more checks in audit, we can do it separately, but for now it's only these 2 simple checks, no need for additional module.

**Comment by mloviska** [comment]
LGTM as smoke test for now. The only thing is the test case header, should be updated and shouldn't the test case check/set `autorelabel` feature?

File: tests/transactional/enable_selinux.pm [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+#
+# Copyright © 2021 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+#
+# Package: transactional-update
+# Summary: Disable Grub timeout by transactional update
```
Comment (b10n1k) [inline tests/transactional/enable_selinux.pm:11]: i know! copy-paste. would you? :)

**Comment by b10n1k** [comment]
> > I wonder if you should add another module which will check for _denial_ in the audit. wdyt?
> 
> If we need to do more checks in audit, we can do it separately, but for now it's only these 2 simple checks, no need for additional module.

those checks, it is likely, that they never fail. and i doubt that anyone will check the uploaded logs. this is my objection. ofc you can merge that but create another ticket to cover that.

**Comment by jlausuch** [comment]
> LGTM as smoke test for now. The only thing is the test case header, should be updated and shouldn't the test case check/set `autorelabel` feature?

Indeed, I missed it and I created a line for it following the [documentation](https://en.opensuse.org/Portal:MicroOS/SELinux). Actually, the boot was failing and there were some errors. Doing another VR to see if this solves the issue. Thanks!

File: tests/transactional/enable_selinux.pm [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+sub run {
+    select_console 'root-console';
+    my $audit_log = "/var/log/audit/audit.log";
+
+    # SELinux is disabled by default in SLE Micro
+    validate_script_output("sestatus", sub { m/SELinux status: .*disabled/ }) if is_sle_micro;
+
+    # Make sure auditd is enabled
+    assert_script_run("systemctl enable auditd");
```
[the remark this answers is not in the document]
Comment (jlausuch) [reply tests/transactional/enable_selinux.pm]: Well, enabling a service that is already enabled doesn't throw errors. This way we avoid an additional step :)

**Review by b10n1k — APPROVED** [review]
+1

File: tests/transactional/enable_selinux.pm (hunk 6 above)
[the remark this answers is not in the document]
Comment (mloviska) [reply tests/transactional/enable_selinux.pm]: According to [systemd presets](https://build.suse.de/package/view_file/SUSE:SLE-15-SP3:Update:Products:MicroOS51/systemd-presets-branding-SMO/50-default-SUSE_MicroOS.preset?expand=1), the service is not enabled by default. Not sure whether that will change in near future, but to start the daemon we need to add `--now` to this call as Yiannis suggests.

**Comment by jlausuch** [comment]
VR:
- [openSUSE MicroOS](http://fromm.arch.suse.de/tests/1185)
- [SLE Micro 5.1](http://fromm.arch.suse.de/tests/1184)  (it fails, but there is a [bug reported](https://bugzilla.suse.com/show_bug.cgi?id=1185412) about this)

File: tests/transactional/enable_selinux.pm (hunk 6 above)
[the remark this answers is not in the document]
Comment (jlausuch) [reply tests/transactional/enable_selinux.pm]: Done. But it's interesting that auditd is enabled and started in sle micro and microos after first boot

File: tests/transactional/enable_selinux.pm [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+use transactional qw(process_reboot);
+use version_utils qw(is_sle_micro);
+use Utils::Systemd qw(systemctl);
+
+sub run {
+    select_console 'root-console';
+    my $audit_log = "/var/log/audit/audit.log";
+
+    # SELinux is disabled by default in SLE Micro
+    validate_script_output("sestatus", sub { m/SELinux status: .*disabled/ }) if is_sle_micro;
```
Comment (lnussel) [inline tests/transactional/enable_selinux.pm:28]: the is_sle_micro shouldn't be needed?

File: tests/transactional/enable_selinux.pm [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+    # Make sure auditd is enabled
+    systemctl 'enable --now auditd';
+
+    assert_script_run('transactional-update setup-selinux');
+    assert_script_run('touch /etc/selinux/.autorelabel');
```
Comment (lnussel) [inline tests/transactional/enable_selinux.pm:34]: doesn't setup-selinux touch that file? If not bug IMO

**Review by lnussel — CHANGES_REQUESTED** [review]

File: tests/transactional/enable_selinux.pm [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+sub run {
+    select_console 'root-console';
+    my $audit_log = "/var/log/audit/audit.log";
+
+    # SELinux is disabled by default in SLE Micro
+    validate_script_output("sestatus", sub { m/SELinux status: .*disabled/ }) if is_sle_micro;
+
+    # Make sure auditd is enabled
+    systemctl 'enable --now auditd';
```
Comment (lnussel) [inline tests/transactional/enable_selinux.pm:31]: this should be the case by default, I think you can assert it's enabled

File: tests/transactional/enable_selinux.pm (hunk 7 above)
Comment (jlausuch) [reply tests/transactional/enable_selinux.pm:28]: selinux is pre-installed in SLE Micro, but not in microOS... so the command will fail there cause there is no such binary.

File: tests/transactional/enable_selinux.pm (hunk 9 above)
Comment (jlausuch) [reply tests/transactional/enable_selinux.pm:31]: You mean throw an error if it's not enabled/running?

Comment (mloviska) [reply tests/transactional/enable_selinux.pm:31]: > You mean throw an error if it's not enabled/running?

And replace the check with `systemctl 'is-enabled auditd`

File: tests/transactional/enable_selinux.pm (hunk 7 above)
Comment (lnussel) [reply tests/transactional/enable_selinux.pm:28]: ah ok in that case I'd use something like `test -d /sys/fs/selinux && test -e /etc/selinux/config`. That's what the `selinuxenabled` command does

File: tests/transactional/enable_selinux.pm (hunk 9 above)
Comment (lnussel) [reply tests/transactional/enable_selinux.pm:31]: yes

File: tests/transactional/enable_selinux.pm (hunk 8 above)
Comment (jlausuch) [reply tests/transactional/enable_selinux.pm:34]: In SLE Micro the file exists cause SELinux is already installed. 
In MicroOS, the file doesn't exist, but after doing `transactional-update setup-selinux` and reboot, there is a `/etc/selinux/.relabelled` file.

**Comment by jlausuch** [comment]
VRs:
[SLE Micro](http://fromm.arch.suse.de/tests/1192)
[MicroOS](http://fromm.arch.suse.de/tests/1193)

File: tests/transactional/enable_selinux.pm [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+    assert_script_run('selinuxenabled');
+    record_info('SELinux',       script_output('sestatus'));
+    record_info('Audit report',  script_output('aureport'));
+    record_info('Audit denials', script_output('aureport -a'));
+    upload_logs($audit_log);
+}
+
+sub test_flags {
+    return {no_rollback => 1};
```
Comment (Vogtinator) [inline tests/transactional/enable_selinux.pm:43]: Shouldn't this be a fatal milestone instead?

File: variables.md [hunk 11]
```diff
@@ -128,6 +128,7 @@ RESET_HOSTNAME| boolean | false | If set to true content of /etc/hostname file w
 SCC_ADDONS | string | | Comma separated list of modules to be enabled using SCC/RMT.
 SCC_DOCKER_IMAGE | string | | The content of /etc/zypp/credentials.d/SCCcredentials used by container-suseconnect-zypp zypper service in SLE base container images
 SELECT_FIRST_DISK | boolean | false | Enables test module to select first disk for the installation. Is used for baremetal machine tests with multiple disks available, including cases when server still has previous installation.
+SELINUX | boolean | false | Enables SELinux tests in transactional server environments.
```
Comment (Vogtinator) [inline variables.md:131]:
```suggestion
SELINUX | boolean | false | Explicitly enable SELinux in transactional server environments.
```

**Comment by Vogtinator** [comment]
Just to be sure: The purpose of the added test is to enable SELinux and not test it, right?
So if at some point SELinux is enabled by default, this test will be skipped?

That makes the `SELINUX` variable pretty misleading, because when installing MicroOS with SELinux as default, `SELINUX=1` will result in a failure...

**Comment by jlausuch** [comment]
> Just to be sure: The purpose of the added test is to enable SELinux and not test it, right?
> So if at some point SELinux is enabled by default, this test will be skipped?
> 
> That makes the `SELINUX` variable pretty misleading, because when installing MicroOS with SELinux as default, `SELINUX=1` will result in a failure...

Yes, that's the goal. I was trying to reuse the code in `./tests/security/selinux`,  but those would need a big refactor to be used for transactional case... We can think of adding more specific selinux tests later on. 

Is MicroOS being to be built with SELinux enabled? If so, you are right and this module will be skipped, but for now this is not the case and it makes sense to execute it to show some errors (as it happens in [SLE Micro case](https://bugzilla.suse.com/show_bug.cgi?id=1185412)).

**Comment by Vogtinator** [comment]
> > Just to be sure: The purpose of the added test is to enable SELinux and not test it, right?
> > So if at some point SELinux is enabled by default, this test will be skipped?
> > That makes the `SELINUX` variable pretty misleading, because when installing MicroOS with SELinux as default, `SELINUX=1` will result in a failure...
> 
> Yes, that's the goal. I was trying to reuse the code in `./tests/security/selinux`, but those would need a big refactor to be used for transactional case... We can think of adding more specific selinux tests later on.

What about renaming the variable then?

> Is MicroOS being to be built with SELinux enabled? If so, you are right and this module will be skipped, but for now this is not the case and it makes sense to execute it to show some errors (as it happens in [SLE Micro case](https://bugzilla.suse.com/show_bug.cgi?id=1185412)).

AFAIK the plan is to make it the default at some point.

**Comment by Vogtinator** [comment]
> > > > Just to be sure: The purpose of the added test is to enable SELinux and not test it, right?
> > > > So if at some point SELinux is enabled by default, this test will be skipped?
> > > > That makes the `SELINUX` variable pretty misleading, because when installing MicroOS with SELinux as default, `SELINUX=1` will result in a failure...
> > > 
> > > 
> > > Yes, that's the goal. I was trying to reuse the code in `./tests/security/selinux`, but those would need a big refactor to be used for transactional case... We can think of adding more specific selinux tests later on.
> > 
> > 
> > What about renaming the variable then?
> > `ENABLE_SELINUX`?

Sounds good.

> > > Is MicroOS being to be built with SELinux enabled? If so, you are right and this module will be skipped, but for now this is not the case and it makes sense to execute it to show some errors (as it happens in [SLE Micro case](https://bugzilla.suse.com/show_bug.cgi?id=1185412)).
> > 
> > 
> > AFAIK the plan is to make it the default at some point.
> > Ok, when that happens we can disable this module, but it should be ok to have it now.

Kind of. Then we have:
a) MicroOS without SELinux by default, kept as-is
b) MicroOS without SELinux by default, enabled manually
c) MicroOS with SELinux by default, kept as-is

For c), would the variable still be used? If not, how would it know that it's the default?

But yeah, can always be changed later when necessary.

**Comment by jlausuch** [comment]
> Kind of. Then we have:
> a) MicroOS without SELinux by default, kept as-is
> b) MicroOS without SELinux by default, enabled manually
> c) MicroOS with SELinux by default, kept as-is
> 
> For c), would the variable still be used? If not, how would it know that it's the default?
No, we would not needed it. We would only test c)

> 
> But yeah, can always be changed later when necessary.

Would you then suggest enabling SELinux in ALL tests by default (now manually, and later on kept as-is)?

**Comment by Vogtinator** [comment]
> > Kind of. Then we have:
> > a) MicroOS without SELinux by default, kept as-is
> > b) MicroOS without SELinux by default, enabled manually
> > c) MicroOS with SELinux by default, kept as-is
> > For c), would the variable still be used? If not, how would it know that it's the default?
> > No, we would not needed it. We would only test c)
> 
> > But yeah, can always be changed later when necessary.
> 
> Would you then suggest enabling SELinux in ALL tests by default (now manually, and later on kept as-is)?

The default configuration is what users are most likely to run, so testing that is the most important setup.

**Comment by jlausuch** [comment]
http://fromm.arch.suse.de/tests/1199
http://fromm.arch.suse.de/tests/1200

**Comment by sysrich** [comment]
> AFAIK the plan is to make it the default at some point.

For openSUSE MicroOS SELinux is already the default and has been for some time - 
https://github.com/yast/skelcd-control-MicroOS/commit/a0c5bb26d2e2fe6969d18855f0510a4aabea58c7

**Comment by jlausuch** [comment]
> > AFAIK the plan is to make it the default at some point.
> 
> For openSUSE MicroOS SELinux is already the default and has been for some time -
> [yast/skelcd-control-MicroOS@a0c5bb2](https://github.com/yast/skelcd-control-MicroOS/commit/a0c5bb26d2e2fe6969d18855f0510a4aabea58c7)

I have tested `Tumbleweed-MicroOS-Image` and it doesn't have selinux enabled by default, not even the packages installed.

**Comment by Vogtinator** [comment]
> > > AFAIK the plan is to make it the default at some point.
> > 
> > 
> > For openSUSE MicroOS SELinux is already the default and has been for some time -
> > [yast/skelcd-control-MicroOS@a0c5bb2](https://github.com/yast/skelcd-control-MicroOS/commit/a0c5bb26d2e2fe6969d18855f0510a4aabea58c7)
> 
> I have tested `Tumbleweed-MicroOS-Image` and it doesn't have selinux enabled by default, not even the packages installed.

The images hit a bug in transactional-update and couldn't be switched yet.

**Comment by jlausuch** [comment]
> > > > AFAIK the plan is to make it the default at some point.
> > > 
> > > 
> > > For openSUSE MicroOS SELinux is already the default and has been for some time -
> > > [yast/skelcd-control-MicroOS@a0c5bb2](https://github.com/yast/skelcd-control-MicroOS/commit/a0c5bb26d2e2fe6969d18855f0510a4aabea58c7)
> > 
> > 
> > I have tested `Tumbleweed-MicroOS-Image` and it doesn't have selinux enabled by default, not even the packages installed.
> 
> The images hit a bug in transactional-update and couldn't be switched yet.

Ok, since this will be the default case for TW, I will enable this only for SLE-Micro.

**Comment by jlausuch** [comment]
2 new runs:
http://fromm.arch.suse.de/tests/1210
http://fromm.arch.suse.de/tests/1211

**Comment by jlausuch** [comment]
@Vogtinator @sysrich ok with the latest change?

**Review by Vogtinator — APPROVED** [review]

**Pull request merged**